overlord/devicemgmtstate/devicemgmtmgr.go

```diff
@@ -0,0 +1,561 @@
+	}
+
+	return nil
+}
+
+// canPoll checks whether polling should happen now based on feature flags and timing.
+func (m *DeviceMgmtManager) canPoll(lastPolled time.Time) bool {
+	tr := config.NewTransaction(m.state)
+	enabled, err := features.Flag(tr, features.MessagePolling)
+	if err != nil && !config.IsNoOption(err) {
+		logger.Noticef("cannot check message-polling feature flag: %v", err)
+		return false
+	}
+	if !enabled {
+		return false
+	}
+
+	if time.Since(lastPolled) < pollingInterval {
+		return false
+	}
+
+	return true
+}
+
+// doPollMessages polls the store for new request messages, acknowledges receipt
+// of persisted messages, and sends queued response messages.
+func (m *DeviceMgmtManager) doPollMessages(t *state.Task, tomb *tomb.Tomb) error {
+	m.state.Lock()
+
+	ms, err := m.getState()
+	if err != nil {
+		return err
+	}
+
+	deviceCtx, err := snapstate.DevicePastSeeding(m.state, nil)
+	if err != nil {
+		m.state.Unlock()
+		return err
+	}
+
+	sto := snapstate.Store(m.state, deviceCtx)
+	m.state.Unlock()
+
+	messages := make([]store.Message, 0, len(ms.ReadyResponses))
+	for _, msg := range ms.ReadyResponses {
+		messages = append(messages, msg)
+	}
+
+	resp, err := sto.PollMessages(
+		tomb.Context(context.Background()),
+		&store.PollMessagesRequest{
+			After:    ms.PendingAckToken,
+			Limit:    pollingLimit,
+			Messages: messages,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	for _, msg := range resp.Messages {
+		pendingMsg, err := parsePendingMessage(msg.Message)
+		if err != nil {
+			logger.Noticef("cannot parse message with token %s: %v", msg.Token, err)
+			continue
+		}
+
+		ms.PendingMessages[pendingMsg.ID()] = pendingMsg
+	}
+
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	ms.ReadyResponses = make(map[string]store.Message)
+	ms.PendingAckToken = ""
+	ms.LastPolled = time.Now()
+
+	if len(resp.Messages) > 0 {
+		ms.PendingAckToken = resp.Messages[len(resp.Messages)-1].Token
+
+		chg := t.Change()
+		var dispatchQueued bool
+		chg.Get("dispatch-queued", &dispatchQueued)
+		if !dispatchQueued {
+			dispatch := m.state.NewTask("dispatch-messages", "Dispatch message(s) to subsystems")
+			dispatch.WaitFor(t)
+			chg.AddTask(dispatch)
+			chg.Set("dispatch-queued", true)
+		}
+	}
+
+	m.setState(ms)
+
+	return nil
+}
+
+// doDispatchMessages selects pending messages for processing & creates validate-message
+// tasks for them.
+// TODO: handle sequencing - pick messages from sequences that have predecessors applied.
+func (m *DeviceMgmtManager) doDispatchMessages(t *state.Task, _ *tomb.Tomb) error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	ms, err := m.getState()
+	if err != nil {
+		return err
+	}
+
+	chg := t.Change()
+	for _, msg := range ms.PendingMessages {
+		if msg.ChangeID != "" {
+			continue
+		}
+
+		validate := m.state.NewTask("validate-message", fmt.Sprintf("Validate message %s", msg.ID()))
+		validate.Set("id", msg.ID())
+		validate.WaitFor(t)
+		chg.AddTask(validate)
```

Comment:
I think we're constructing the change very incrementally and it might make it tough to follow/debug. We could do it in a two stage process: first we poll for new messages/send responses, then based on that we create the rest.
We should be able to create the validate, apply and queue-response tasks and pass data between them through the state instead of baking it into the task on creation. For instance, if the validation failed we could have the `apply` be a no-op and the `queue-response` could skip waiting and build the rejection immediately. Does that seem workable?